File: lib/medium/remove_duplicate_letters.dart
String removeDuplicateLetters(String s) {
  Map<String, int> last = {};
  List<String> stack = [];
  Set<String> seen = {};
  for (int i = 0; i < s.length; i++) {
    last[s[i]] = i;
  }
  for (var i = 0; i < s.length; i++) {
    if (seen.contains(s[i])) continue;
    while (stack.isNotEmpty && stack.last.compareTo(s[i]) > 0 && (last[stack.last] ?? -1) > i) {
      seen.remove(stack.removeLast());
    }
    stack.add(s[i]);
    seen.add(s[i]);
  }
  return stack.join("");
}

File: medium/remove_duplicate_letters.go
package medium

func RemoveDuplicateLetters(s string) string {
	last := make(map[rune]int)
	seen := make(map[rune]bool)
	stack := make([]rune, 0, len(s))

	for i, char := range s {
		last[char] = i
	}

	for i, char := range s {
		if seen[char] {
			continue
		}

		for len(stack) > 0 {
			top := stack[len(stack)-1]
			if top > char && last[top] > i {
				stack = stack[:len(stack)-1]
				delete(seen, top) // ✅ correct
			} else {
				break
			}
		}

		stack = append(stack, char)
		seen[char] = true
	}

	return string(stack)
}
